src/proc/rotation-filter.cpp

```diff
@@ -0,0 +1,214 @@
+                break;
+            }
+                
+            default:
+                LOG_ERROR( "Rotation transform does not support format: "
+                           + std::string( rs2_format_to_string( tgt.get_profile().format() ) ) );
+            }
+            return tgt;
+        }
+        return f;
+    }
+
+    void  rotation_filter::update_output_profile(const rs2::frame& f)
+    {
+        _source_stream_profile = f.get_profile();
+        
+        _target_stream_profile = _source_stream_profile.clone( _source_stream_profile.stream_type(),
+                                                            _source_stream_profile.stream_index(),
+                                                            _source_stream_profile.format() );
+
+
+        auto src_vspi = dynamic_cast< video_stream_profile_interface * >( _source_stream_profile.get()->profile );
+        if( ! src_vspi )
+            throw std::runtime_error( "Stream profile interface is not video stream profile interface" );
+
+        auto tgt_vspi = dynamic_cast< video_stream_profile_interface * >( _target_stream_profile.get()->profile );
+        if( ! tgt_vspi )
+            throw std::runtime_error( "Profile is not video stream profile" );
+
+        rs2_intrinsics src_intrin = src_vspi->get_intrinsics();
+        rs2_intrinsics tgt_intrin = tgt_vspi->get_intrinsics();
+
+        // Adjust width and height based on the rotation angle 
+        if( _value == 90 || _value == -90 )  // 90 or -90 degrees rotation
+        {
+            _rotated_width = src_intrin.height;
+            _rotated_height = src_intrin.width;
+            tgt_intrin.fx = src_intrin.fy;
+            tgt_intrin.fy = src_intrin.fx;
+            tgt_intrin.ppx = src_intrin.ppy;
+            tgt_intrin.ppy = src_intrin.ppx;
+        }
+        else if( _value == 180 )  // 180 degrees rotation
+        {
+            _rotated_width = src_intrin.width;
+            _rotated_height = src_intrin.height;
+            tgt_intrin = src_intrin;
+        }
+        else { throw std::invalid_argument( "Unsupported rotation angle" ); }
+
+        tgt_intrin.width = _rotated_width;
+        tgt_intrin.height = _rotated_height;
+
+        tgt_vspi->set_intrinsics( [tgt_intrin]() { return tgt_intrin; } );
+        tgt_vspi->set_dims( _rotated_width, _rotated_height );
+    }
+
+    rs2::frame rotation_filter::prepare_target_frame(const rs2::frame& f, const rs2::frame_source& source, rs2_extension tgt_type)
+    {
+        auto vf = f.as<rs2::video_frame>();
+        auto ret = source.allocate_video_frame(_target_stream_profile, f,
+            vf.get_bytes_per_pixel(),
+            _rotated_width,
+            _rotated_height,
+            _rotated_width * vf.get_bytes_per_pixel(),
+            tgt_type);
+
+        return ret;
+    }
+
+    template< size_t SIZE >
+    void rotation_filter::rotate_depth( uint8_t * const out, const uint8_t * source,
```

Comment:
Please explore some improvements ChatGPT offers,
Could be small but worth exploring:

1. Calculate Repeated Values Outside the Loops
2. Reduce Branching by Using Separate Functions for Each Rotation
3. Minimize std::memcpy Calls
4. Parallel Processing (if openMP is available)

```
void rotation_filter::rotate_depth(uint8_t* const out, const uint8_t* source, int width, int height) {
    auto width_out = (_value == 90 || _value == -90) ? height : width;
    auto height_out = (_value == 90 || _value == -90) ? width : height;

    switch (_value) {
        case 90:
            rotate_90(out, source, width, height, width_out, height_out);
            break;
        case -90:
            rotate_neg90(out, source, width, height, width_out, height_out);
            break;
        case 180:
            rotate_180(out, source, width, height, width_out, height_out);
            break;
        default:
            throw std::invalid_argument("Unsupported rotation angle");
    }
}

void rotation_filter::rotate_90(uint8_t* const out, const uint8_t* source, int width, int height, int width_out, int height_out) {
    for (int i = 0; i < height; ++i) {
        int reverse_i = width_out - i - 1;
        for (int j = 0; j < width; ++j) {
            auto src_index = (i * width + j) * SIZE;
            auto out_index = (j * width_out + reverse_i) * SIZE;
            if (SIZE == 1) {
                out[out_index] = source[src_index];
            } else {
                std::memcpy(&out[out_index], &source[src_index], SIZE);
            }
        }
    }
}

// Define rotate_neg90 and rotate_180 similarly
```

Reply:
I'm not sure it will be better, we will have 3 almost identical functions (rotate_90, rotate 180, rotate -90) changing only the out_index

Reply:
Your call..
What about 1,3,4?

Reply:
Fixed